python/ray/ray_cluster_perf.py

```diff
@@ -39,6 +70,21 @@ def f(object_ref_list):
     print("latency to get an 1G object over network", round(time_diff, 2),
           "+-", round(time_diff_std, 2))
 
+    arr1 = np.ones((40, 1024, 1024))
+    arr2 = np.ones((40, 1024, 1024))
+
+    def data_intensive_task():
+        ray.get(large_arg_and_value.remote(arr1, arr2))
```

Comment:
I don't think this actually tests data locality, does it? The arr1 and arr2 variables will get captured and shipped with the remote function, I think. Probably a better way is to return an array from a task, then pass it to another task.

But, actually I'm not sure if we really want these benchmarks here. It seems more useful to just have the Python tests and maybe a larger-scale release test that runs on physical nodes.

Reply:
>I don't think this actually tests data locality, does it?

No, I don't think so! My guess is that both argument objects will be created locally and the task will therefore be scheduled locally, the same as no locality-aware leasing. I was mainly hoping to show that there are no performance regressions in this trivial case and planning on constructing benchmarks that test data locality via worker resource constraints + task chains and fan-ins.

>The arr1 and arr2 variables will get captured and shipped with the remote function, I think.

Hmm how so? `arr1` and `arr2` aren't captured within scope of the `large_arg_and_value` remote function since that's defined outside of the `main` function, while the former is defined within the `main` function, so `arr1` and `arr2` shouldn't get shipped on function export. Shouldn't these ndarrays be serialized and sent as args on each `large_arg_and_value.remote()` invocation? Or are you talking about how the arguments are always put into Plasma locally and therefore the task will always be scheduled locally, same as no locality-aware leasing?

>Probably a better way is to return an array from a task, then pass it to another task.

I was going to try out some large chains of tasks (that won't reuse a lease) anyway in order to more reliably trigger locality-aware leasing, although I was still planning on feeding a plain ndarray argument to the first task. What are the advantages to having the source data be a task return object instead of a plain serialized argument? I suppose if that source data task is scheduled on a different worker, we'd reach a nontrival data locality decision faster than if the source data was a plain serialized argument (which is always put into Plasma locally)?

>But, actually I'm not sure if we really want these benchmarks here.

Agreed, but what is `ray_cluster_perf` used for, given that `ray clusterbenchmark` isn't exposed anymore? Should I remove these kind of data-intensive microbenchmarks from `ray_perf` and keep them here, or should I create a separate module to house them? I was hoping to not add clutter to the Python module space so I dropped them in here. 😁 

>It seems more useful to just have the Python tests and maybe a larger-scale release test that runs on physical nodes.

Definitely. Is the CI infrastructure set up somewhere for an auto-deployed physical cluster for running these kind of benchmarks? Like a GCP instance group that only the CI system has permission to deploy to and run benchmarks on? Because that would be pretty cool!

Assuming that doesn't already exist, in the meantime, I could save some benchmarks that properly test locality-aware leasing in a Python module that can be run on any cluster. The cluster will just have to be manually set up.

Reply:
@stephanie-wang I'll remove these microbenchmarks from `ray_perf.py` and `ray_cluster_perf.py`, and if I end up creating some benchmarks that properly hit locality-aware leasing, I'll put those in a separate module that only has the task definitions and some helper code to run and measure, i.e. doesn't simulate a cluster (it would expect a cluster to have already been manually set up). Does that sound good?

Reply:
Thanks, that sounds good! Yes, so I think to merge:
1. Remove microbenchmarks.
2. Add a Python test (you could put it in test_advanced_3.py) that checks that the data locality scheduling is working. Basically the same thing that you have under the cluster benchmarks right now, but just checking system behavior, not performance.

About the way the test case is written, I meant for you to pass an ObjectRef directly to another task. Large objects that get captured in the remote function's scope won't show up in the task's dependencies because they're part of the function description. So the data locality scheduling that you added should only apply for explicit task arguments.

Reply:
That sounds good!

>Large objects that get captured in the remote function's scope won't show up in the task's dependencies because they're part of the function description. So the data locality scheduling that you added should only apply for explicit task arguments.

But those large arguments `arr1` and `arr2` are defined within `main()`, outside of `large_arg_and_value()`'s scope (which is defined at the module level, not inside of `main()`), so how would those large objects get captured in that remote function's function description? I might be missing something.

Reply:
Ah I see, I misread the original test and thought they were getting captured in the remote function's definition. Ignore that part!

But yes, I do think the test could be improved by returning the arguments from a remote task instead of putting the plasma objects locally, like you mentioned.

Reply:
Ok cool, just making sure y'all didn't recently add any scope capture jitting on remote function invocation that I wasn't aware of! 😁 

Definitely agree, I'll make sure to do that in the Python test(s).